File: server/src/entities/task/Task.ts
import {
    Column,
    CreateDateColumn,
    Entity,
    Index,
    JoinColumn,
    JoinTable,
    ManyToMany,
    ManyToOne,
    OneToMany,
    PrimaryGeneratedColumn,
    UpdateDateColumn
} from "typeorm"
import { Comment } from "../comment/Comment"
import { Project } from "../project/Project"
import { User } from "../user/User"
import { TaskPriority } from "./PriorityEnum"
import { TaskStatus } from "./StatusEnum"
import { Tag } from "./Tag"

@Entity("taskr_tasks")
export class Task {
    @PrimaryGeneratedColumn("uuid")
    id: string

    @Column({ unique: true, nullable: false })
    @Index()
    title: string

    @Column({ nullable: false, type: "text" })
    description: string

    @ManyToOne(() => Project, (project) => project.tasks, { eager: false })
    @JoinColumn({ name: "projectId" })
    project: Project

    @ManyToOne(() => User, (user) => user.tasks, { eager: false })
    @JoinColumn({ name: "creatorId" })
    creator: User

    @Column({ type: "enum", enum: TaskStatus, default: TaskStatus.PLANNED })
    status: TaskStatus

    @Column({ type: "enum", enum: TaskPriority, default: TaskPriority.LOW })
    priority: TaskPriority

    @Column({ nullable: false, type: "timestamp" })
    dueDate: Date

    @OneToMany(() => Comment, (comment) => comment.task, {
        cascade: true,
        lazy: true,
        onDelete: "CASCADE",
        onUpdate: "CASCADE"
    })
    comments?: Promise<Comment[]>

    @ManyToMany(() => User, (user) => user.assignedTasks, { lazy: true })
    @JoinTable({
        name: "taskr_task_assignees",
        joinColumn: { name: "taskId", referencedColumnName: "id" },
        inverseJoinColumn: { name: "userId", referencedColumnName: "id" }
    })
    assignees?: Promise<User[]>

    @ManyToMany(() => Tag, (tag) => tag.tasks, { cascade: true, eager: true })
    @JoinTable({
        name: "taskr_tasks_tags",
        joinColumn: { name: "taskId" },
        inverseJoinColumn: { name: "tagId" }
    })
    tags?: Promise<Tag[]>

    @CreateDateColumn({ type: "timestamp", default: () => "CURRENT_TIMESTAMP" })
    createdAt: Date

    @UpdateDateColumn({ type: "timestamp", default: () => "CURRENT_TIMESTAMP", onUpdate: "CURRENT_TIMESTAMP" })
    updatedAt: Date
}

File: server_springboot_demo/src/main/java/com/barataribeiro/taskr/entities/tasks/Task.java
package com.barataribeiro.taskr.entities.tasks;

import java.time.LocalDate;
import java.util.Date;
import java.util.List;
import java.util.UUID;

import org.hibernate.annotations.ColumnDefault;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;

import com.barataribeiro.taskr.entities.comment.Comment;
import com.barataribeiro.taskr.entities.project.Project;
import com.barataribeiro.taskr.entities.user.User;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.JoinTable;
import jakarta.persistence.ManyToMany;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.Table;
import jakarta.persistence.Temporal;
import jakarta.persistence.TemporalType;
import lombok.Data;

@Data
@Entity
@Table(name = "taskr_tasks")
public class Task {

  @Id
  @GeneratedValue(strategy = GenerationType.UUID)
  private UUID id;

  private String title;

  private String description;

  @ManyToOne(fetch = FetchType.LAZY)
  @JoinColumn(name = "projectId")
  private List<Project> projects;

  @ManyToOne(fetch = FetchType.LAZY)
  @JoinColumn(name = "creatorId")
  private List<User> creator;

  @ColumnDefault("'PLANNED'")
  private StatusEnum status;

  @ColumnDefault("'LOW'")
  private PriorityEnum priority;

  @Temporal(TemporalType.TIMESTAMP)
  @Column(nullable = false)
  private LocalDate dueDate;

  @OneToMany(fetch = FetchType.LAZY, mappedBy = "task", cascade = CascadeType.ALL)
  private List<Comment> comments;

  @ManyToMany(fetch = FetchType.LAZY)
  @JoinTable(name = "taskr_tasks_assignees", joinColumns = @JoinColumn(name = "taskId"), inverseJoinColumns = @JoinColumn(name = "userId"))
  private List<User> assignees;

  @ManyToMany(fetch = FetchType.LAZY)
  @JoinTable(name = "taskr_tasks_tags", joinColumns = @JoinColumn(name = "taskId"), inverseJoinColumns = @JoinColumn(name = "tagId"))
  private List<Tag> tags;

  @CreatedDate
  @Temporal(TemporalType.TIMESTAMP)
  @Column(name = "created_at", nullable = false, updatable = false)
  private Date createdAt;

  @LastModifiedDate
  @Temporal(TemporalType.TIMESTAMP)
  @Column(name = "updated_at")
  private Date updatedAt;

}
